### ISSUE TITLE
Basic support for creating entities on partitioned tables in SQL Server (Code-First)

### ISSUE THREAD

**Issue by airbreather** [dotnet/efcore#33505]
## Background

SQL Server supports something called [partitioned tables](https://learn.microsoft.com/en-us/sql/relational-databases/partitions/create-partitioned-tables-and-indexes). In one of our ORM-free legacy applications, we've used this functionality to great effect to accelerate managing large data sets with [minimal logging](https://learn.microsoft.com/en-us/sql/relational-databases/import-export/prerequisites-for-minimal-logging-in-bulk-import), after our first attempt at these kinds of tools backfired when the transaction log file tried to outgrow the disk it was on.

## What problem are you trying to solve?

Trying to replicate our solution in another application, which uses EF Core (Code-First style), has been a challenge. With many other SQL Server features, even if EF Core doesn't support them directly, `migrationBuilder.Sql("...");` is good enough...

...but not for partitioned tables. I can't just create a table (or index) as normal and then issue a command that says "take this existing table (or index) and put it onto that partition scheme". The table (or index) needs to be **created on** that partition scheme, so the entire `CREATE TABLE` or (`CREATE INDEX`) would need to go into `migrationBuilder.Sql("...");`.

My first thought was to inherit from `SqlServerMigrationsSqlGenerator` and override `Generate(CreateTableOperation, ...)` to tweak the `CREATE TABLE` statement to add `ON ThePartitionScheme(ThePartitioningColumn)` to the right spot (given the right custom annotations), but this was incomplete: `RelationalDatabaseCreator.EnsureCreatedAsync` tries to run those very migration operations* when the database is created, and so I also needed my `Generate(CreateTableOperation, ...)` override to be able to start by creating the partition function and scheme if those do not exist.
- **despite [the documentation for `DatabaseFacade.EnsureCreatedAsync`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.databasefacade.ensurecreatedasync) emphatically claiming that it doesn't use migrations*

Ultimately, I have something that **I think** is going to be functional **for a while**, but I'm worried how much custom code I had to write in order to do this, including a suppression or two for `EF1001`.

## Describe the solution you'd like

I don't think my use case needs particularly rich support in order to see a huge improvement. I think that I would be happy if I could write something like this in my `DbContext`:

```csharp
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    // include the full SQL to create the partition function and scheme in their initial states.
    modelBuilder
        .HasPartitionFunction(createSql: "CREATE PARTITION FUNCTION PF_Blog /* ... */")
        .HasPartitionScheme(createSql: "CREATE PARTITION SCHEME PS_Blog /* ... */");

    // entity types themselves should have an easier time.
    modelBuilder.Entity<Blog>()
        .HasPartitioning(partitionScheme: "PS_Blog", partitionColumn: "BlogId");
}
```

I'm going to keep this initial post short and add more context, thoughts, and details into a comment.

**Comment by airbreather** [comment]
> Since the function and scheme are defined as SQL in any case, HasPartitionFunction/Scheme really provide little advantage over simply introducing the appropriate CREATE PARTITION directly into a migration.

If not for the wrinkle about `RelationalDatabaseCreator.EnsureCreatedAsync` running the SQL generated by `Generate(CreateTableOperation, ...)`, I would completely agree.

As it stands, from my perspective (I've got to run right now, so I can't elaborate), it **looks like** there's no way to use `RelationalDatabaseCreator.EnsureCreatedAsync` in an application that **also** wants to use partitioned tables?

**Comment by roji** [comment]
Yeah, EnsureCreatedAsync is indeed incompatible with custom SQL, and that's (only) one reason why it's generally unsuitable for real applications. It really is a feature for test code and demos, nothing more - so yeah, I don't think there's a way to use EnsureCreatedAsync and have partitioned tables (or any other DDL/migration feature that EF doesn't handle).

**Comment by airbreather** [comment]
That makes sense. For context, we use it to initially create a LocalDB database for automated tests that we immediately wipe out with a `RESTORE DATABASE` command since our data seeding code is a bit slow to do the whole thing on every single CI build. `MigrateAsync` comes after that, since we don't want to have to recreate that baseline database backup file for every single migration no matter how small...

I can just switch that `EnsureCreatedAsync` call to another `MigrateAsync`, though, if that's the recommended alternative to adding spot for running arbitrary SQL between when the database is created and when the tables are created.

**Comment by roji** [comment]
Yep, test database management is indeed a whole domain that can be pretty complex :)

So yeah, I'd definitely recommend switching to MigrateAsync for your test database creation; assuming you do use migrations for your production app, this also guarantees that your tests run on the same database schema that your actual app does (notably around questions of custom SQL). The main possible disadvantage of that is that if you start to have a big migration history, that history gets replayed when creating the test database; a possible mitigation is to squash your migration history every now and then and start over. Though if you're creating your database only once when starting a CI run (which is generally good), that hopefully isn't a big deal.

Data seeding specifically is something that should generally be handled separately (especially if lots of data is involved), and so should be orthogonal to how you create your database or schema (i.e. EnsureCreated or Migrate).

**Comment by yedajiang44** [comment]
When I wanted to use efcore and pgsql to implement a partition table, I read a lot of information and finally found that there was no support (#https://github.com/npgsql/efcore.pg/issues/1035).

 I am a little overwhelmed now. I hope efcore can support it.

**Comment by jscarle** [comment]
I just painted myself into a corner without realizing it. I never had a need for table partitioning before, so it never came into considering while designing a particular service we recently spun up. Now the service is running in production, built around EF Core (as are all of our services), and this particular service has become host to a very large table that holds 22 million rows that include 10 years of historical data.

Investigations have led me down the road that table partitioning (per year, for example) would solve many of our performance issues. Though it's now obvious that I've hit a roadblock going down that path with EF Core. I'll surely find my way down to another solution, but I just wanted to add a +1 that this would have been nice to have support for.

**Comment by roji** [comment]
@jscarle and others, AFAIK partitioning is importantly relevant only within the context of creating/managing your table (DDL); in other words, this is a migrations feature that doesn't affect querying, inserting, etc. As such, you can perfectly well use EF to query and update a SQL Server partitioned table - the only missing piece is that EF won't create such tables for you.

EF generally doesn't support all the DDL possibilities of SQL Server (and all other database) - that's unfeasible, and generally doesn't provide great value; but users can (and are encouraged to) customize their migrations to set up table partitioning or any other feature they desire. In other words, you really haven't painted yourself into any corner by using EF - you simply have to manually perform the creation of the table in question, and then continue using EF as before.

**Comment by jscarle** [comment]
Thanks for the clarity on this @roji.

I ended up solving this by going a slightly different route that made it easier to maintain by preserving the out-of-the-box migrations functionality. I essentially used an approach of database sharding by having two identical DbContexts, one that's connected to one database with the current and recent data, the other connected to the other database with the historical data. Since both use the same entities and configurations, migrations are done on both in parallel and they both maintain the same schema. A wrote a simple background service that periodically moves data out of the current database to the historical one by instantiating both DbContexts and passing the data from one to the other. A bit more I/O and infrastructure work that partitioning, but this allows EF migrations to continue handling the DDL tasks.

**Comment by roji** [comment]
> I ended up solving this by going a slightly different route that made it easier to maintain by preserving the out-of-the-box migrations functionality.

It's great that you managed to find a workable solution. But I'll point out that you can absolutely continue using EF's migrations even when you need some unsupported DDL feature (like partitioning) - it's only the specific operation (creation of the partitioned table in this case) that must be done manually. Beyond that operation, all other migrations can be done as usual.

In other words, if SQL Server table partitioning is something that matches your needs, you absolutely should not be avoiding it just because you want to use EF's migrations - this simply isn't an all-or-nothing thing. Use migrations for everything you can, and drop down to SQL customization of migrations where that's not possible.

**Comment by jscarle** [comment]
What I've found, in my experience, is the more you stray away from things that differ from what EF understands in its internal model, the more work that translates into when creating migrations as those become additional things to remember and adjust for manually.

I've run into several situations where I've refactored code in ways that are "EF friendly" versus what "Azure SQL understands" to avoid such additional overhead.

One recent example, unrelated to partitioning, was dealing with indexes that include properties of an owned type (see #32037). Can be it adjusted for manually? Sure, but it can be avoided alltogether by not using an owned entity on that particular column.

I suppose in the end, it a questions of choosing how much you're willing to manually adjust and fine-tune your database versus just letting EF manage it.

**Comment by ErikEJ** [comment]
Or use Database model first and make your life much simpler??

**Comment by jscarle** [comment]
> Or use Database model first and make your life much simpler??

That's subjective. I'd even argue that people that prefer to manually adjust their database model would instead likely be using Dapper.

I feel that most of us who use Entity Framework do so because we'd prefer to focus on the code and not concern ourselves as much on the database aspect of it. Now that doesn't mean that database configuration isn't important, but I think it's obvious by the comments made by @roji in what I believe was the [.NET Data Community Standup for EF Core 9](https://www.youtube.com/watch?v=wG8D5HJMzjA) that even though there are things we should not be doing a certain way, we still continue to do them for simplicity's sake. Running migrations on application startup comes to mind as an example, and though it's well known that we shouldn't, many of us still continue to this day, and this eventually led to the introduction of [migrations locks](https://github.com/dotnet/efcore/issues/33731).

But I digress, the original point I was making is simply that although there's always a work around, many of us still find value in having more support brought into EF for these types of things.

**Comment by adriaan-booysen-bt** [comment]
We have done a library for the company I work for that supports partitioning of SQL server and EntityFramework, their requirement was for Guid's, although support was build in for Guid, numeric, dates and various forms (can't remember all off them), it uses EF API's to build the partitions, the implementation after partitioning is only to add the following to the configuration, below is a sample of a table (for testing) and the a constant was used for the schema/partition name

```
    internal class ShortEntityConfiguration : BaseTypeConfiguration<ShortEntity, short>
    {
        public override void Configure(EntityTypeBuilder<ShortEntity> builder)
        {
            base.Configure(builder);
            builder.ToTable(nameof(ShortEntity)).OnPartition(Constants.ShortSchemeName, x => x.Value);
        }
    }
```

I do have permission to include this as a PR for further contribution to the community ... this is being used in an application where the production SQL database is >20TB and has billions of rows in Azure ...

We want to make this a community effort/project and was thinking of starting an EntityFramework{Core}.SqlServer.Partitioning namespace for this ...

@roji we did chat about this in the 2024 Techorama in Belgium, if the community is keen on this, we will gladly contribute our initial work

We will gladly be willing to discuss items, although support would be limited.

Please upvote with a 👍if you are interested

**Comment by roji** [comment]
Thanks @adriaan-booysen-bt! Depending on the extent of the feature and what it looks like, this may make sense to be included in the SQL Server provider itself, rather than as an extension. However, up to now it doesn't seem like we've seen much community requests for this (this issue has only a single vote).

Let's follow this and see.
